elastic/indexer/insta_users/main.go

```diff
@@ -0,0 +1,71 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/go-elasticsearch/v7"
+
+	esIndexer "github.com/codeuniversity/smag-mvp/elastic/indexer"
+	"github.com/codeuniversity/smag-mvp/kafka/changestream"
+	"github.com/codeuniversity/smag-mvp/service"
+	"github.com/codeuniversity/smag-mvp/utils"
+)
+
+const esIndex = "insta_user"
+
+func main() {
+	kafkaAddress := utils.GetStringFromEnvWithDefault("KAFKA_ADDRESS", "my-kafka:9092")
+	groupID := utils.GetStringFromEnvWithDefault("KAFKA_GROUPID", "insta_usersearch-inserter")
+	changesTopic := utils.GetStringFromEnvWithDefault("KAFKA_CHANGE_TOPIC", "postgres.public.users")
+
+	esHosts := utils.GetMultipliesStringsFromEnvDefault("ELASTIC_SEARCH_ADDRESS", []string{"http://localhost:9201"})
+
+	// create and run esInserter
+	i := esIndexer.New(esHosts, esIndex, instaUserMapping, kafkaAddress, changesTopic, groupID, handleChangemessage)
+
+	service.CloseOnSignal(i)
+	waitUntilClosed := i.Start()
+
+	waitUntilClosed()
+}
+
+// handleChangemessage filters relevant events and upserts them
+func handleChangemessage(esClient *elasticsearch.Client, m *changestream.ChangeMessage) error {
+	user := &user{}
+	if err := json.Unmarshal(m.Payload.After, user); err != nil {
+		return err
+	}
+
+	switch m.Payload.Op {
+	case "c", "r", "u":
+		return upsertDocument(user, esClient)
+	}
+
+	return nil
+}
+
+func upsertDocument(u *user, esClient *elasticsearch.Client) error {
+	// not using esapi.NewJSONReader, because we will wrap the user into the upsert request
+	jsonUser, err := json.Marshal(*u)
+	if err != nil {
+		return err
+	}
+
+	searchUser := fmt.Sprintf(instaUserUpsert, jsonUser, jsonUser)
```

Comment:
`fmt.Sprintf(` to modify the upsert json directly is dangerous this opens us up to malicious content injections - instead we should define a for the request and marshal the whole struct (unfortunately, this is a bit annoying to do) 
`+` this doesn't guarantee that we generate well formed, valid json 
`+` this is not the best style to generate json in general 😅